[Stats] Metrics Export User Interface Part 1 (#9913)
ray-project/ray: doc/source/ray-metrics.rst

Monitoring Ray
==============
To help monitoring Ray applications, Ray

- Collects Ray's pre-selected system level metrics.
- Exposes metrics through Prometheus endpoints.
- Supports application-level metrics collection APIs.

.. note::

    It is currently an experimental feature and under active development. APIs are subject to change.

Getting Started (Single Node)
-----------------------------
You can access Ray's metrics through Prometheus endpoints.

First, let's learn how to access metrics when you use `ray.init()`.

.. code-block:: python

    import ray

>     ray.init(_metrics_export_port=8080)

Reviewer: What's the reason for the underscore? Because it's experimental?

    import time
    time.sleep(100)

Now you can access Prometheus endpoint at the address `http://localhost:8080`. Note that if you don't specify `_metrics_export_port`, 

> the port is randomly assigned. You can check the port number when you run `ray.init`.

Reviewer: Should we do the same thing we do with the dashboard? This seems like a bad UX

Author: I think it is fine because users will eventually just need to run Prometheus with json files for service discovery. I'd love to change it if there's any user request in the future though.

.. code-block:: bash

    {'node_ip_address': '192.168.1.82',
    'raylet_ip_address': '192.168.1.82',
    'redis_address': '192.168.1.82:6379',
    'object_store_address': '/tmp/ray/session_2020-08-04_13-50-35_156166_20127/sockets/plasma_store',
    'raylet_socket_name': '/tmp/ray/session_2020-08-04_13-50-35_156166_20127/sockets/raylet',
    'webui_url': 'localhost:8265',
    'session_dir': '/tmp/ray/session_2020-08-04_13-50-35_156166_20127',
    'metrics_export_port': 63301} # <= You can access metrics through http://localhost:[metrics_export_port]

Second, let's learn how to expose metrics when you use `ray start`.

.. code-block:: bash

    ray start --head --metrics-export-port=8080 # Assign metrics export port on a head node.

Now, you can scrape Ray's metrics using Prometheus. 

> First, download Prometheus.

Reviewer: You should say it's exported in prometheus format, can be scraped by anything but here we'll use prometheus for simplicity

We will walkthrough how to import metrics from a Ray cluster using Prometheus.

At a head node,

.. code-block:: bash

    ray start --head --metrics-export-port=8080 # Assign metrics export port on a head node.

At a worker node,

.. code-block:: bash

    ray start --address=[head_node_address] --metrics-export-port=8080

You can now get the url of metrics agents using `ray.nodes()`

.. code-block:: python

    import ray
    ray.init(address='auto')
    from pprint import pprint
    pprint(ray.nodes())

    """
    [{'Alive': True,
    'MetricsExportPort': 8080,
    'NodeID': '02ec0a33d3e75a520fd07238e9363b48a442ce24',
    'NodeManagerAddress': '192.168.1.82',
    'NodeManagerHostname': 'host-MBP.attlocal.net',
    'NodeManagerPort': 57167,
    'ObjectManagerPort': 56201,
    'ObjectStoreSocketName': '/tmp/ray/session_2020-08-04_13-57-29_819087_20360/sockets/plasma_store',
    'RayletSocketName': '/tmp/ray/session_2020-08-04_13-57-29_819087_20360/sockets/raylet',
    'Resources': {'CPU': 16.0,
                    'memory': 91.0,
                    'node:192.168.1.82': 1.0,
                    'object_store_memory': 31.0},
    'alive': True}]
    """

> Now, setup your prometheus to read metrics from `[NodeManagerAddress]:[MetricsExportPort]` from all nodes in the cluster.

Reviewer: Link to the prometheus service discovery page

Monitoring Ray

> ==============

Reviewer 2: Using Ray and Prometheus
========================

Author: Maybe `Ray Monitoring with Prometheus`? (I think it is important to add a keyword monitoring)

> To help monitoring Ray applications, Ray

Reviewer 2: This page describes ...

- Collects Ray's pre-selected system level metrics.
- Exposes metrics through Prometheus endpoints.

> - Supports application-level metrics collection APIs (EXPERIMENTAL).

Reviewer 2: This allows you to ...

> - Supports application-level metrics collection APIs (EXPERIMENTAL).

Author: Removed application level metrics information for now cc @edoakes

.. note::

    It is currently an experimental feature and under active development. APIs are subject to change.

> Getting Started (Single Node)

Reviewer 2: Can we just drop the `ray.init()` documentation for now? 

The UX seems quite odd (it seems like you're exposing an internal configuration via the underscore, and it doesn't seem like a common use case).

-----------------------------

> Ray exposes its metrics in Prometheus format. For the simplicty, let's use Prometheus to scrape these metrics.

Reviewer 2: ```suggestion
Ray exposes its metrics in Prometheus format. This allows us to ... 

For simplicity, let's walk through an example to use Prometheus to scrape these metrics.
```

> First, let's learn how to access metrics when you use `ray.init()`.

Reviewer 2: ```suggestion
First, let's expose metrics via `ray.init()`.
```

Author: This section is removed.

.. code-block:: python

    import ray
    ray.init(_metrics_export_port=8080)
    import time
    time.sleep(100)

Now you can access Prometheus endpoint at the address `http://localhost:8080`. Note that if you don't specify `_metrics_export_port`, 
the port is randomly assigned. You can check the port number when you run `ray.init`.

.. code-block:: bash

    {'node_ip_address': '192.168.1.82',
    'raylet_ip_address': '192.168.1.82',
    'redis_address': '192.168.1.82:6379',
    'object_store_address': '/tmp/ray/session_2020-08-04_13-50-35_156166_20127/sockets/plasma_store',
    'raylet_socket_name': '/tmp/ray/session_2020-08-04_13-50-35_156166_20127/sockets/raylet',
    'webui_url': 'localhost:8265',
    'session_dir': '/tmp/ray/session_2020-08-04_13-50-35_156166_20127',
    'metrics_export_port': 63301} # <= You can access metrics through http://localhost:[metrics_export_port]

> Second, let's learn how to expose metrics when you use `ray start`.

Reviewer 2: ```suggestion
Second, let's expose metrics through `ray start`.
```

.. code-block:: bash

    ray start --head --metrics-export-port=8080 # Assign metrics export port on a head node.

Now, you can scrape Ray's metrics using Prometheus. 

First, download Prometheus.

.. code-block:: bash

    tar xvfz prometheus-*.tar.gz

>     cd prometheus-*

Reviewer 2: I think you should provide a downloadable link? Otherwise, this code doesn't seem particularly informative.

Author: Good catch. I will add it.

Let's modify Prometheus's config file to scrape metrics from Prometheus endpoints.

.. code-block:: yaml

    # prometheus.yml
    global:
    scrape_interval:     5s
    evaluation_interval: 5s

    rule_files:
    # - "first.rules"
    # - "second.rules"

    scrape_configs:
    - job_name: prometheus
        static_configs:
        - targets: ['localhost:8080'] # This must be same as metrics_export_port

Next, let's start Prometheus.

.. code-block:: shell

    ./prometheus --config.file=./prometheus.yml

Now, you can access Ray metrics from the default Prometheus url, `http://localhost:9090`.

Getting Started (Multi-nodes)
-----------------------------

> Ray runs a metrics agent per node. Each metrics agent collects metrics from a local node and exposes it through a Prometheus endpoint.
> 
> We will walkthrough how to import metrics from a Ray cluster using Prometheus.

Reviewer 2: ```suggestion

Let's now walk through how to import metrics from a Ray cluster.

Ray runs a metrics agent per node. Each metrics agent collects metrics from a local node and exposes it through a Prometheus endpoint.

You can then use each endpoint to... [?]
```

.. code-block:: bash

    ray start --head --metrics-export-port=8080 # Assign metrics export port on a head node.

At a worker node,

.. code-block:: bash

    ray start --address=[head_node_address] --metrics-export-port=8080

You can now get the url of metrics agents using `ray.nodes()`

.. code-block:: python

    import ray
    ray.init(address='auto')
    from pprint import pprint
    pprint(ray.nodes())

    """
    [{'Alive': True,
    'MetricsExportPort': 8080,
    'NodeID': '02ec0a33d3e75a520fd07238e9363b48a442ce24',
    'NodeManagerAddress': '192.168.1.82',
    'NodeManagerHostname': 'host-MBP.attlocal.net',
    'NodeManagerPort': 57167,
    'ObjectManagerPort': 56201,
    'ObjectStoreSocketName': '/tmp/ray/session_2020-08-04_13-57-29_819087_20360/sockets/plasma_store',
    'RayletSocketName': '/tmp/ray/session_2020-08-04_13-57-29_819087_20360/sockets/raylet',
    'Resources': {'CPU': 16.0,
                    'memory': 91.0,
                    'node:192.168.1.82': 1.0,
                    'object_store_memory': 31.0},
    'alive': True}]
    """

Now, setup your prometheus to read metrics from `[NodeManagerAddress]:[MetricsExportPort]` from all nodes in the cluster.

> If you'd like to make this process automated, you can also use `file based service discovery <https://prometheus.io/docs/guides/file-sd/#installing-configuring-and-running-prometheus>`_.

Reviewer 2: What would this do?

Author: Added additional explanation.
